File: test-fixtures/languages/java/complex/AdvancedPatterns.java
package com.bank.api;

import java.sql.*;
import java.io.*;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.*;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class AdvancedPatterns {

    private Connection connection;

    public AdvancedPatterns(Connection connection) {
        this.connection = connection;
    }

    // ============================================================
    // VULN 1: Stream API chain flowing taint to SQL
    // Chain: request.getParameter() -> stream().map().filter().collect()
    //        -> String.join() -> Statement.executeQuery()
    // The taint flows through the Stream pipeline's map/filter/collect
    // operations without sanitization, ending up in a raw SQL query.
    // ============================================================

    public void searchByTags(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String tagsParam = request.getParameter("tags");
        String[] rawTags = tagsParam.split(",");

        String whereClause = Arrays.stream(rawTags)
            .map(tag -> tag.trim())
            .filter(tag -> !tag.isEmpty())
            .map(tag -> "tag = '" + tag + "'")
            .collect(Collectors.joining(" OR "));

        String sql = "SELECT * FROM accounts WHERE " + whereClause;
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    // ============================================================
    // VULN 2: Optional chain preserving taint
    // Chain: request.getParameter() -> Optional.ofNullable()
    //        -> .map() -> .orElse() -> Statement.executeQuery()
    // The Optional wrapper does not sanitize; .map() transforms
    // the tainted value into a SQL fragment, and .orElse() unwraps it.
    // ============================================================

    public void findAccountByAlias(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String alias = request.getParameter("alias");

        String condition = Optional.ofNullable(alias)
            .map(a -> "alias = '" + a + "'")
            .orElse("1=0");

        String sql = "SELECT * FROM accounts WHERE " + condition;
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    // ============================================================
    // VULN 3: Lambda as argument to method
    // Chain: request.getParameter() -> processWithTransform() lambda arg
    //        -> transform.apply() -> Statement.executeQuery()
    // The tainted value is passed to a higher-order method that
    // applies a caller-supplied lambda, which builds unsafe SQL.
    // ============================================================

    public void searchWithTransform(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String input = request.getParameter("search");

        String result = processWithTransform(input, value -> "name LIKE '%" + value + "%'");

        String sql = "SELECT * FROM accounts WHERE " + result;
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    private String processWithTransform(String value, Function<String, String> transform) {
        if (value == null) {
            return "1=1";
        }
        return transform.apply(value);
    }

    // ============================================================
    // VULN 4: Method reference (Class::method)
    // Chain: request.getParameter() -> List.of() -> stream()
    //        -> .map(AdvancedPatterns::toSqlLiteral) -> collect()
    //        -> Statement.executeQuery()
    // The method reference AdvancedPatterns::toSqlLiteral wraps
    // each tainted element into a SQL literal without escaping.
    // ============================================================

    public void searchByIds(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String idsParam = request.getParameter("ids");
        List<String> ids = List.of(idsParam.split(","));

        String inClause = ids.stream()
            .map(AdvancedPatterns::toSqlLiteral)
            .collect(Collectors.joining(", "));

        String sql = "SELECT * FROM accounts WHERE id IN (" + inClause + ")";
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    private static String toSqlLiteral(String value) {
        return "'" + value.trim() + "'";
    }

    // ============================================================
    // VULN 5: Try-with-resources with tainted path
    // Chain: request.getParameter() -> new File() -> new FileInputStream()
    //        -> try-with-resources block -> response output
    // The user-controlled filename is used directly to construct a
    // File path, enabling path traversal attacks.
    // ============================================================

    public void readDocument(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String docName = request.getParameter("doc");
        String basePath = "/var/bank/documents/";
        File target = new File(basePath + docName);

        try (FileInputStream fis = new FileInputStream(target);
             BufferedInputStream bis = new BufferedInputStream(fis)) {
            byte[] buffer = new byte[4096];
            int bytesRead;
            while ((bytesRead = bis.read(buffer)) != -1) {
                response.getOutputStream().write(buffer, 0, bytesRead);
            }
        }
    }

    // ============================================================
    // VULN 6: instanceof pattern matching (Java 16+)
    // Chain: request.getParameter() -> parseInput() returns Object
    //        -> instanceof String s -> s used in SQL
    // The instanceof pattern variable `s` captures the tainted
    // string, which then flows unsanitized into a SQL statement.
    // ============================================================

    public void handleDynamicInput(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String raw = request.getParameter("filter");
        Object parsed = parseInput(raw);

        if (parsed instanceof String s) {
            String sql = "SELECT * FROM accounts WHERE " + s;
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery(sql);
            writeResults(response, rs);
        }
    }

    private Object parseInput(String raw) {
        if (raw == null) {
            return Integer.valueOf(0);
        }
        return raw;
    }

    // ============================================================
    // VULN 7: Switch expression (Java 14+)
    // Chain: request.getParameter("type") -> switch expression
    //        -> each branch builds SQL with request.getParameter("q")
    //        -> Statement.executeQuery()
    // The switch expression selects which SQL template to use, but
    // all branches incorporate the tainted query parameter.
    // ============================================================

    public void queryByType(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String type = request.getParameter("type");
        String queryValue = request.getParameter("q");

        String sql = switch (type) {
            case "name" -> "SELECT * FROM accounts WHERE name = '" + queryValue + "'";
            case "email" -> "SELECT * FROM accounts WHERE email = '" + queryValue + "'";
            case "status" -> "SELECT * FROM accounts WHERE status = '" + queryValue + "'";
            default -> "SELECT * FROM accounts WHERE id = '" + queryValue + "'";
        };

        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    // ============================================================
    // VULN 8: Record class storing tainted data
    // Chain: request.getParameter() -> new AuditEntry(userId, action)
    //        -> record accessor entry.action() -> SQL concatenation
    //        -> Statement.executeQuery()
    // The record stores the tainted value and its accessor returns
    // it without sanitization, flowing into a raw SQL query.
    // ============================================================

    public record AuditEntry(String userId, String action, long timestamp) {
        public String toWhereClause() {
            return "user_id = '" + userId + "' AND action = '" + action + "'";
        }
    }

    public void searchAuditLog(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String userId = request.getParameter("userId");
        String action = request.getParameter("action");
        AuditEntry entry = new AuditEntry(userId, action, System.currentTimeMillis());

        String clause = entry.toWhereClause();
        String sql = "SELECT * FROM audit_log WHERE " + clause;
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    // ============================================================
    // VULN 9: StringBuilder fluent append chain to SQL
    // Chain: request.getParameter() -> StringBuilder.append() chain
    //        -> .toString() -> Statement.executeQuery()
    // Multiple tainted parameters are appended into a StringBuilder
    // via fluent method chaining, building an injectable SQL query.
    // ============================================================

    public void advancedSearch(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String name = request.getParameter("name");
        String city = request.getParameter("city");
        String minBalance = request.getParameter("minBalance");

        String sql = new StringBuilder("SELECT * FROM accounts WHERE 1=1")
            .append(" AND name LIKE '%").append(name).append("%'")
            .append(" AND city = '").append(city).append("'")
            .append(" AND balance >= ").append(minBalance)
            .toString();

        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    // ============================================================
    // VULN 10: CompletableFuture chain
    // Chain: request.getParameter() -> CompletableFuture.supplyAsync()
    //        -> .thenApply() builds SQL -> .thenApply() executes query
    // The tainted value flows through the async pipeline: the first
    // stage wraps it, the second builds SQL, the third executes it.
    // ============================================================

    public void asyncSearch(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String keyword = request.getParameter("keyword");

        CompletableFuture<String> future = CompletableFuture.supplyAsync(() -> keyword)
            .thenApply(k -> "SELECT * FROM accounts WHERE name LIKE '%" + k + "%'");

        String sql = future.get(5, TimeUnit.SECONDS);
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    // ============================================================
    // VULN 11: Custom @FunctionalInterface with command injection
    // Chain: request.getParameter() -> CommandBuilder.build()
    //        -> Runtime.exec()
    // A custom functional interface is used to build a command
    // string from tainted input, then executed via Runtime.exec().
    // ============================================================

    @FunctionalInterface
    interface CommandBuilder {
        String build(String input);
    }

    public void exportData(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String format = request.getParameter("format");

        CommandBuilder builder = fmt -> "data-export --format " + fmt + " --output /tmp/export";
        String command = builder.build(format);
        Process process = Runtime.getRuntime().exec(command);

        try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
            String line;
            StringBuilder output = new StringBuilder();
            while ((line = reader.readLine()) != null) {
                output.append(line);
            }
            response.getWriter().write(output.toString());
        }
    }

    // ============================================================
    // VULN 12: Stream reduce accumulating tainted fragments
    // Chain: request.getParameter() -> split -> stream().reduce()
    //        -> accumulated SQL string -> Statement.executeQuery()
    // The stream reduce operation concatenates tainted fragments
    // into a single SQL WHERE clause without sanitization.
    // ============================================================

    public void filterByMultiple(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String filtersParam = request.getParameter("filters");
        String[] parts = filtersParam.split(";");

        String whereClause = Arrays.stream(parts)
            .reduce("1=1", (acc, part) -> acc + " AND " + part);

        String sql = "SELECT * FROM accounts WHERE " + whereClause;
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        writeResults(response, rs);
    }

    // ============================================================
    // VULN 13: ProcessBuilder with tainted arguments list
    // Chain: request.getParameter() -> List.of() -> new ProcessBuilder()
    //        -> .start()
    // User-controlled values are placed into a ProcessBuilder
    // argument list, enabling command injection.
    // ============================================================

    public void runDiagnostic(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String host = request.getParameter("host");
        String port = request.getParameter("port");

        List<String> command = new ArrayList<>();
        command.add("nmap");
        command.add("-p");
        command.add(port);
        command.add(host);

        ProcessBuilder pb = new ProcessBuilder(command);
        Process process = pb.start();

        try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
            String line;
            StringBuilder output = new StringBuilder();
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
            response.getWriter().write(output.toString());
        }
    }

    // ============================================================
    // Helpers
    // ============================================================

    private void writeResults(HttpServletResponse response, ResultSet rs) throws Exception {
        response.setContentType("application/json");
        PrintWriter out = response.getWriter();
        ResultSetMetaData meta = rs.getMetaData();
        int cols = meta.getColumnCount();
        out.write("[");
        boolean first = true;
        while (rs.next()) {
            if (!first) out.write(",");
            out.write("{");
            for (int i = 1; i <= cols; i++) {
                if (i > 1) out.write(",");
                out.write("\"" + meta.getColumnName(i) + "\":\"" + rs.getString(i) + "\"");
            }
            out.write("}");
            first = false;
        }
        out.write("]");
    }

    // ============================================================
    // ADDITIONAL PATTERNS - Annotation-based dispatch, inheritance
    // ============================================================

    // Annotation processing (simulated - actual annotations need framework)
    // @RequestMapping("/api/exec")
    public String annotatedHandler(String userInput) throws Exception {
        // Annotation marks this as HTTP handler entry point
        return Runtime.getRuntime().exec(userInput).toString();
    }

    // Inheritance chain with virtual dispatch
    abstract static class BaseHandler {
        abstract String handle(String data);
        String process(String data) {
            return handle(data);  // virtual dispatch to subclass
        }
    }

    static class SqlHandler extends BaseHandler {
        @Override
        String handle(String data) {
            // Override reaches SQL sink
            return "SELECT * FROM t WHERE id = " + data;
        }
    }

    // Switch expression (Java 14+)
    static String switchExpression(String type, String payload) {
        return switch (type) {
            case "exec" -> payload;  // taint flows through switch arm
            case "safe" -> "sanitized";
            default -> throw new IllegalArgumentException(type);
        };
    }

    // Try-with-resources flow
    static void tryWithResources(String query) throws Exception {
        try (var conn = java.sql.DriverManager.getConnection("jdbc:h2:mem:test");
             var stmt = conn.createStatement()) {
            stmt.executeQuery(query);  // taint through try-with-resources
        }
    }

    // Varargs propagation
    static void varargsSink(String... commands) {
        for (String cmd : commands) {
            try { Runtime.getRuntime().exec(cmd); } catch (Exception e) {}
        }
    }

    // Builder pattern with taint
    static class QueryBuilder {
        private StringBuilder query = new StringBuilder();
        QueryBuilder select(String cols) { query.append("SELECT ").append(cols); return this; }
        QueryBuilder from(String table) { query.append(" FROM ").append(table); return this; }
        QueryBuilder where(String cond) { query.append(" WHERE ").append(cond); return this; }
        String build() { return query.toString(); }
    }
}

    // Enum with behavior
    enum CommandType {
        EXEC("exec"),
        QUERY("query");
        final String value;
        CommandType(String v) { this.value = v; }
    }

File: test-fixtures/languages/typescript/complex/advanced_patterns.ts
import { Request, Response } from 'express';
import * as fs from 'fs';
import { execSync } from 'child_process';

// ---------------------------------------------------------------------------
// Types and interfaces
// ---------------------------------------------------------------------------

interface DatabaseClient {
    query(sql: string, params?: any[]): Promise<any[]>;
}

interface Identifiable {
    id: string;
}

interface HasPayload {
    payload: string;
}

// Discriminated union for event routing
type AppEvent =
    | { kind: 'user_command'; command: string }
    | { kind: 'file_upload'; filename: string; content: string }
    | { kind: 'query_request'; sql: string }
    | { kind: 'eval_script'; code: string };

// Enum for report format routing
enum ReportFormat {
    SQL = 'sql',
    Shell = 'shell',
    Template = 'template',
    File = 'file',
}

// Interface for dispatch pattern
interface DataProcessor {
    process(data: string): Promise<string>;
}

// Mapped type alias
type FieldRecord = Record<string, string>;

// ---------------------------------------------------------------------------
// VULN 1: Generic function with type parameter flowing taint
// Chain: req.query.filter -> extractField<T>() -> db.query(unsanitized)
// The generic preserves the tainted string through the type parameter.
// ---------------------------------------------------------------------------

function extractField<T extends Identifiable & HasPayload>(
    obj: T,
    db: DatabaseClient
): Promise<any[]> {
    const value: string = obj.payload;
    const sql: string = "SELECT * FROM items WHERE data = '" + value + "'";
    return db.query(sql);
}

async function handleGenericRoute(req: Request, res: Response, db: DatabaseClient): Promise<void> {
    const item: Identifiable & HasPayload = {
        id: req.query.id as string,
        payload: req.query.filter as string,
    };
    const results = await extractField(item, db);
    res.json(results);
}

// ---------------------------------------------------------------------------
// VULN 2: Type guard narrowing -- taint survives the `is` check
// Chain: req.body.input -> isStringInput() narrows type -> db.query(tainted)
// The type guard confirms the shape but does NOT sanitize the value.
// ---------------------------------------------------------------------------

interface StringInput {
    kind: 'string';
    value: string;
}

interface NumberInput {
    kind: 'number';
    value: number;
}

type MixedInput = StringInput | NumberInput;

function isStringInput(input: MixedInput): input is StringInput {
    return input.kind === 'string';
}

async function handleTypeGuardRoute(req: Request, res: Response, db: DatabaseClient): Promise<void> {
    const input: MixedInput = req.body.input;
    if (isStringInput(input)) {
        // Type guard confirms shape, but value is still user-controlled
        const sql: string = "SELECT * FROM data WHERE col = '" + input.value + "'";
        const rows = await db.query(sql);
        res.json(rows);
    } else {
        res.json({ value: input.value });
    }
}

// ---------------------------------------------------------------------------
// VULN 3: Discriminated union switch routing tainted data to sinks
// Chain: req.body -> AppEvent discriminated union -> switch(kind) -> multiple sinks
// Each branch of the switch sends tainted data to a different dangerous sink.
// ---------------------------------------------------------------------------

async function dispatchEvent(event: AppEvent, db: DatabaseClient): Promise<string> {
    switch (event.kind) {
        case 'user_command': {
            // VULN 3a: command injection
            const output: string = execSync('run-task ' + event.command).toString();
            return output;
        }
        case 'file_upload': {
            // VULN 3b: path traversal + arbitrary file write
            fs.writeFileSync('/uploads/' + event.filename, event.content);
            return 'uploaded';
        }
        case 'query_request': {
            // VULN 3c: SQL injection
            const rows = await db.query("SELECT * FROM logs WHERE " + event.sql);
            return JSON.stringify(rows);
        }
        case 'eval_script': {
            // VULN 3d: code injection
            const result: any = eval(event.code);
            return String(result);
        }
    }
}

async function handleEventRoute(req: Request, res: Response, db: DatabaseClient): Promise<void> {
    const event: AppEvent = req.body as AppEvent;
    const result: string = await dispatchEvent(event, db);
    res.send(result);
}

// ---------------------------------------------------------------------------
// VULN 4: Async generator yielding tainted values to a sink
// Chain: req.query.terms -> streamSearch() async generator -> db.query(tainted)
// The async generator yields tainted SQL results without sanitization.
// ---------------------------------------------------------------------------

async function* streamSearch(
    terms: string[],
    db: DatabaseClient
): AsyncGenerator<any[], void, unknown> {
    for (const term of terms) {
        const sql: string = "SELECT * FROM products WHERE name LIKE '%" + term + "%'";
        const batch: any[] = await db.query(sql);
        yield batch;
    }
}

async function handleStreamRoute(req: Request, res: Response, db: DatabaseClient): Promise<void> {
    const terms: string[] = (req.query.terms as string).split(',');
    const allResults: any[] = [];
    for await (const batch of streamSearch(terms, db)) {
        allResults.push(...batch);
    }
    res.json(allResults);
}

// ---------------------------------------------------------------------------
// VULN 5: Interface dispatch -- taint flows through interface method
// Chain: req.body.data -> SqlProcessor.process() -> db.query(tainted)
// The interface type hides the concrete implementation that sinks to SQL.
// ---------------------------------------------------------------------------

class SqlProcessor implements DataProcessor {
    private db: DatabaseClient;

    constructor(db: DatabaseClient) {
        this.db = db;
    }

    async process(data: string): Promise<string> {
        const sql: string = "INSERT INTO audit_log (entry) VALUES ('" + data + "')";
        await this.db.query(sql);
        return 'logged';
    }
}

async function handleInterfaceRoute(req: Request, res: Response, db: DatabaseClient): Promise<void> {
    const processor: DataProcessor = new SqlProcessor(db);
    const input: string = req.body.data;
    const result: string = await processor.process(input);
    res.json({ status: result });
}

// ---------------------------------------------------------------------------
// VULN 6: Mapped type / Record with tainted values
// Chain: req.body.fields (Record) -> buildInsert() -> db.query(tainted)
// Tainted values inside a Record<string,string> flow into SQL concatenation.
// ---------------------------------------------------------------------------

function buildInsert(table: string, fields: FieldRecord): string {
    const columns: string = Object.keys(fields).join(', ');
    const values: string = Object.values(fields).map(v => "'" + v + "'").join(', ');
    return `INSERT INTO ${table} (${columns}) VALUES (${values})`;
}

async function handleMappedTypeRoute(req: Request, res: Response, db: DatabaseClient): Promise<void> {
    const fields: FieldRecord = req.body.fields;
    const sql: string = buildInsert('user_profiles', fields);
    await db.query(sql);
    res.json({ status: 'inserted' });
}

// ---------------------------------------------------------------------------
// VULN 7: Decorator on method -- taint flows through decorated method
// Chain: req.params.id -> AdminController.deleteItem() -> exec(tainted)
// The decorator wraps the method but does not sanitize arguments.
// ---------------------------------------------------------------------------

function logExecution(target: any, key: string, descriptor: PropertyDescriptor): PropertyDescriptor {
    const original = descriptor.value;
    descriptor.value = function (...args: any[]) {
        console.log(`Calling ${key} with`, args);
        return original.apply(this, args);
    };
    return descriptor;
}

class AdminController {
    @logExecution
    runCleanup(targetPath: string): string {
        // VULN: command injection -- decorator does not sanitize
        const output: string = execSync('rm -rf ' + targetPath).toString();
        return output;
    }
}

async function handleDecoratorRoute(req: Request, res: Response): Promise<void> {
    const controller = new AdminController();
    const targetPath: string = req.params.path;
    const result: string = controller.runCleanup(targetPath);
    res.send(result);
}

// ---------------------------------------------------------------------------
// VULN 8: Enum switch routing taint to different sinks
// Chain: req.body.format (enum) + req.body.content -> switch -> multiple sinks
// The enum selects which dangerous sink receives the tainted content.
// ---------------------------------------------------------------------------

async function generateReport(
    format: ReportFormat,
    content: string,
    db: DatabaseClient
): Promise<string> {
    switch (format) {
        case ReportFormat.SQL: {
            const sql: string = "INSERT INTO reports (body) VALUES ('" + content + "')";
            await db.query(sql);
            return 'stored in db';
        }
        case ReportFormat.Shell: {
            const output: string = execSync('echo ' + content + ' | format-report').toString();
            return output;
        }
        case ReportFormat.Template: {
            const result: any = eval('`' + content + '`');
            return String(result);
        }
        case ReportFormat.File: {
            fs.writeFileSync('/tmp/report_' + Date.now() + '.txt', content);
            return 'written to file';
        }
    }
}

async function handleEnumRoute(req: Request, res: Response, db: DatabaseClient): Promise<void> {
    const format: ReportFormat = req.body.format as ReportFormat;
    const content: string = req.body.content;
    const result: string = await generateReport(format, content, db);
    res.json({ result });
}

// ---------------------------------------------------------------------------
// VULN 9: Assertion function -- asserts x is T -- taint survives assertion
// Chain: req.body.config -> assertIsConfig() -> exec(tainted field)
// The assertion function validates structure but the string value is still tainted.
// ---------------------------------------------------------------------------

interface ServerConfig {
    host: string;
    command: string;
    port: number;
}

function assertIsConfig(val: unknown): asserts val is ServerConfig {
    if (typeof val !== 'object' || val === null) {
        throw new Error('Not an object');
    }
    const obj = val as Record<string, unknown>;
    if (typeof obj.host !== 'string' || typeof obj.command !== 'string' || typeof obj.port !== 'number') {
        throw new Error('Invalid config shape');
    }
}

async function handleAssertionRoute(req: Request, res: Response): Promise<void> {
    const raw: unknown = req.body.config;
    assertIsConfig(raw);
    // After assertion, TypeScript knows raw is ServerConfig, but values are still user-controlled
    const output: string = execSync(raw.command).toString();
    res.json({ host: raw.host, output });
}

// ---------------------------------------------------------------------------
// VULN 10: Nullish coalesce assignment (??=) with tainted fallback
// Chain: req.query.snippet -> cachedSnippets[key] ??= taintedValue -> eval(cached)
// The ??= operator assigns the tainted value when the cache is empty,
// then the cached (tainted) value is passed to eval.
// ---------------------------------------------------------------------------

const cachedSnippets: Record<string, string> = {};

async function handleNullishCoalesceRoute(req: Request, res: Response): Promise<void> {
    const key: string = req.query.key as string;
    const snippet: string = req.query.snippet as string;
    cachedSnippets[key] ??= snippet;
    const result: any = eval(cachedSnippets[key]);
    res.json({ result });
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

export {
    extractField,
    handleGenericRoute,
    isStringInput,
    handleTypeGuardRoute,
    dispatchEvent,
    handleEventRoute,
    streamSearch,
    handleStreamRoute,
    SqlProcessor,
    handleInterfaceRoute,
    buildInsert,
    handleMappedTypeRoute,
    AdminController,
    handleDecoratorRoute,
    generateReport,
    handleEnumRoute,
    assertIsConfig,
    handleAssertionRoute,
    handleNullishCoalesceRoute,
    ReportFormat,
    AppEvent,
    DataProcessor,
    FieldRecord,
    ServerConfig,
};

// Optional chaining flow
function optionalChainFlow(req?: { query?: { cmd?: string } }): void {
    const cmd = req?.query?.cmd;
    if (cmd) {
        exec(cmd);
    }
}
